64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+#include "nbl/application_templates/MonoDeviceApplication.hpp"
+#include "nbl/application_templates/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace video;
+
+// A high level overview of the algorithm implementation (merge sort).
+// This will be a X pass algorithm, where X is log2(number_of_elements).
+// In each phase, you have to 'merge' 2 sorted list of elements into a single list.
+// Each of these 'lists' will have thier number of elements roughly doubled in each phase (after the merge step of previous phase).
+class MergeSortApp final : public nbl::application_templates::MonoDeviceApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
+{
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+
+public:
+	MergeSortApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Initialize base classes.
+		if (!device_base_t::onAppInitialized(smart_refctd_ptr(system)))
+		{
+			return false;
+		}
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+		{
+			return false;
+		}
+
+		// Get the maxinum bytes of shared memory we can use.
+		// In case the number of elements shared memory array can hold exceeds the limit, then switch to fetching data from global memory directly (for as few memory elements as possible).
+		const auto physicalDeviceLimits = m_device->getPhysicalDevice()->getLimits();
+		const uint32_t maxNumberOfArrayElementsSharedMemoryCanHold = physicalDeviceLimits.maxComputeSharedMemorySize / sizeof(int);
```

Comment:
> I've used shared memory as well. Basically, I put as much data as possible into shared memory. In case the input data is large and more than the physical devices shared memory limit, I fetch as little data as possible from global memory.

You get optimum residency while using about 16 to 32 bytes of smem per workgroup invocation, just make that a tweakable parameter.

If you end up using MAX then you end up with a occupancy problem (see CUDA occupancy calculator)